lib/std/rand/Xoshiro256.zig

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2015-2021 Zig Contributors
+// This file is part of [zig](https://ziglang.org/), which is MIT licensed.
+// The MIT license requires this copyright notice to be included in all copies
+// and substantial portions of the software.
+
+//! Xoshiro256++ - http://xoroshiro.di.unimi.it/
+//!
+//! PRNG
+
+const std = @import("std");
+const Random = std.rand.Random;
+const math = std.math;
+const Xoshiro256 = @This();
+
+random: Random,
+
+s: [4]u64,
+
+pub fn init(init_s: u64) Xoshiro256 {
+    var x = Xoshiro256{
+        .random = Random{ .fillFn = fill },
+        .s = undefined,
+    };
+
+    x.seed(init_s);
+    return x;
+}
+
+fn next(self: *Xoshiro256) u64 {
+    const r = math.rotl(u64, self.s[0] +% self.s[3], 23) +% self.s[0];
+
+    const t = self.s[1] << 17;
+
+    self.s[2] ^= self.s[0];
+    self.s[3] ^= self.s[1];
+    self.s[1] ^= self.s[2];
+    self.s[0] ^= self.s[3];
+
+    self.s[2] ^= t;
+
+    self.s[3] = math.rotl(u64, self.s[3], 45);
+
+    return r;
+}
+
+// Skip 2^128 places ahead in the sequence
+fn jump(self: *Xoshiro256) void {
+    var s: u256 = 0;
+
+    var table: u256 = 0x39abdc4529b1661ca9582618e03fc9aad5a61266f0c9392c180ec6d33cfd0aba;
+
+    while (table != 0) : (table >>= 1) {
+        if (@truncate(u1, table) != 0) {
+            s ^= @bitCast(u256, self.s);
```

Comment:
This assumes that `u256` and `u64` have the same alignment.

Maybe it would be safer to define `s` as `[4]u64` and avoid the bitcasts.  This is something the compiler should be able to vectorize anyway.

Reply:
I'm not sure if I understood the problem correctly : Does `@bitCast` require both type to have same alignment?

Reply:
It doesn't. Sorry for the confusion. The only requirements is that types have the same size.

Avoiding the bitcasts seems to produce slightly better code, though, avoiding unnecessary loads: https://godbolt.org/z/heMEao4Kv